tensornetwork/network_components_test.py

```diff
@@ -0,0 +1,426 @@
+    assert edge.axis1 == 1
+    assert edge.node2 == node2
+    assert edge.axis2 == 1
+    assert edge._is_dangling is False
+    assert edge.signature == -1
+
+
+def test_edge_initialize_raises_error_faulty_arguments(double_node_edge):
+    node1 = double_node_edge.node1
+    node2 = double_node_edge.node2
+    with pytest.raises(ValueError):
+        Edge(name="edge", node1=node1, node2=node2, axis1=0)
+    with pytest.raises(ValueError):
+        Edge(name="edge", node1=node1, axis1=0, axis2=0)
+
+
+def test_edge_set_signature(double_node_edge):
+    edge = double_node_edge.edge12
+    edge.set_signature(2)
+    assert edge.signature == 2
+
+
+def test_edge_set_signature_raises_error_dangling(single_node_edge):
+    edge = single_node_edge.edge
+    with pytest.raises(ValueError):
+        edge.set_signature(2)
+
+
+def test_edge_get_nodes_single(single_node_edge):
+    node = single_node_edge.node
+    edge = single_node_edge.edge
+    assert edge.get_nodes() == [node, None]
+
+
+def test_edge_get_nodes_double(double_node_edge):
+    node1 = double_node_edge.node1
+    node2 = double_node_edge.node2
+    edge = double_node_edge.edge12
+    assert edge.get_nodes() == [node1, node2]
+
+
+def test_edge_update_axis(double_node_edge):
+    node1 = double_node_edge.node1
+    node2 = double_node_edge.node2
+    edge = double_node_edge.edge1
+    edge.update_axis(old_axis=0, old_node=node1, new_axis=2, new_node=node2)
+    assert edge.node1 == node2
+    assert edge.axis1 == 2
+
+
+def test_edge_update_axis_raises_error_old_node(double_node_edge):
+    node2 = double_node_edge.node2
+    edge = double_node_edge.edge1
+    with pytest.raises(ValueError):
+        edge.update_axis(old_axis=0, old_node=node2, new_axis=2, new_node=node2)
+
+
+def test_edge_node1_property(single_node_edge):
+    node = single_node_edge.node
+    edge = single_node_edge.edge
+    assert edge.node1 == node
+
+
+def test_edge_node1_setter(double_node_edge):
+    node2 = double_node_edge.node2
+    edge = double_node_edge.edge1
+    edge.node1 = node2
+    assert edge.node1 == node2
+
+
+def test_edge_node2_property(double_node_edge):
+    node2 = double_node_edge.node2
+    edge1 = double_node_edge.edge1
+    edge12 = double_node_edge.edge12
+    assert edge1.node2 is None
+    assert edge12.node2 == node2
+
+
+def test_edge_node2_setter(double_node_edge):
+    node1 = double_node_edge.node1
+    edge12 = double_node_edge.edge12
+    edge12.node2 = node1
+    assert edge12.node2 == node1
+
+
+def test_edge_dimension(single_node_edge):
+    edge = single_node_edge.edge
+    assert edge.dimension == 1
+
+
+def test_edge_is_dangling(double_node_edge):
+    edge1 = double_node_edge.edge1
+    edge12 = double_node_edge.edge12
+    assert edge1.is_dangling()
+    assert not edge12.is_dangling()
+
+
+def test_edge_is_trace_true(single_node_edge):
+    node = single_node_edge.node
+    edge = Edge(name="edge", node1=node, axis1=2, node2=node, axis2=2)
```

Comment:
We should never  have an edge point the the same axis twice one the same node. Can you change the axis value here? (We should also raise an error if this happens, but we can do that in a later PR)